### PR TITLE
[Frontend] Support complex message content for chat completions endpoint

### PR DESCRIPTION
**Pull request by fgreinacher**
The vLLM OpenAI server currently does not support complex message contents for the [chat completions endpoint](https://platform.openai.com/docs/api-reference/chat/create):

```
    "messages": [
      {
        "role": "user",
        "content": [
          {
            "type": "text",
            "text": "What’s in this image?"
          },
          {
            "type": "image_url",
            "image_url": {
              "url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
            }
          }
        ]
      }
    ]
```

It seems non-trivial/impossible to fully support this format because it strongly depends on the active model.

What we can support easily though are simple cases where just a simple text content is provided:

```
    "messages": [
      {
        "role": "user",
        "content": [
          {
            "type": "text",
            "text": "What is LiteLLM?"
          }
        ]
      }
    ]
```

This does not seem super useful, but it helps for cases where a client library passes complex contents by default, even if they could be represented by a simple string, for example https://github.com/OkGoDoIt/OpenAI-API-dotnet.

:hammer_and_pick: with :heart: by [Siemens](https://opensource.siemens.com/)

<details>

<summary><b> PR Checklist (Click to expand. Please read before submitting.) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

---

Please provide a brief explanation of the motivation behind the PR and the changes it introduces. This helps reviewers understand the context and rationale for the contribution. If possible, please link existing issues this PR will resolve.

### CODE REVIEW

**Comment by fgreinacher** [comment]
This is now ready for review.

There is failure in the [Kernels test](https://buildkite.com/vllm/ci/builds/2966#018e5630-0557-4bb9-90a1-d2258cb50965), but it does not seem to be related to my changes.

File: tests/entrypoints/test_openai_server.py [hunk 1]
```diff
@@ -676,6 +676,22 @@ async def test_response_format_json_object(server, client: openai.AsyncOpenAI):
...
+                                                    "user",
+                                                    "content": [{
+                                                        "type":
+                                                        "text",
+                                                        "text":
+                                                        "what is 1+1?"
+                                                    }]
+                                                }])
+    content = resp.choices[0].message.content
+    assert "2" in content
```
Comment (LiuXiaoxuanPKU) [inline tests/entrypoints/test_openai_server.py:692]: `ground_truth = ....`
Can we directly `content == ground_truth`?

File: vllm/entrypoints/openai/protocol.py [hunk 2]
```diff
@@ -60,10 +68,31 @@ class ResponseFormat(BaseModel):
...
 
+class TextContent(BaseModel):
+    type: str = Literal["text"]
+    text: str
+
+
+SingleTextContentList = Annotated[List[TextContent], Len(1, 1)]
+
+
+def try_flatten_text_content(value: any) -> str:
```
Comment (LiuXiaoxuanPKU) [inline vllm/entrypoints/openai/protocol.py:79]:
```suggestion
def try_flatten_text_content(value: Any) -> str:
```

**Review by LiuXiaoxuanPKU — APPROVED** [review]
Thanks for the contribution! Just some minor comments.

File: vllm/entrypoints/openai/protocol.py [hunk 3]
```diff
@@ -60,10 +68,31 @@ class ResponseFormat(BaseModel):
...
+    # The OpenAI API allows complex contents in chat messages. While
+    # we do not fully support them at the moment, we can allow simple
+    # cases where just a single text content is passed in a message.
+    with contextlib.suppress(ValidationError):
```
Comment (LiuXiaoxuanPKU) [inline vllm/entrypoints/openai/protocol.py:83]: I am a bit concerned here, the validator is triggered for each request. The validator is strict and contains multiple steps. I'm not sure how much overhead will it introduce. Any comments or thoughts? @simon-mo @fgreinacher

Comment (fgreinacher) [reply vllm/entrypoints/openai/protocol.py:83]: I don't expect this to be a problem. Validation happens anyway for every request and this just does a bit more. But we could optimize a bit by skipping this validation if `value` is a plain string.

File: tests/entrypoints/test_openai_server.py (hunk 1 above)
Comment (fgreinacher) [reply tests/entrypoints/test_openai_server.py:692]: Not sure, the model will probably give slightly different answers. It's also not really important, we mostly want to ensure that the request is accepted.

File: vllm/entrypoints/openai/protocol.py (hunk 3 above)
Comment (fgreinacher) [reply vllm/entrypoints/openai/protocol.py]: I have pushed that optimization. With that performance will be unchanged for simple text-only requests.

Comment (LiuXiaoxuanPKU) [reply vllm/entrypoints/openai/protocol.py]: Yeah looks good to me now. Please just fix the conflict and I will merge it, thanks!

Comment (fgreinacher) [reply vllm/entrypoints/openai/protocol.py]: Thanks @LiuXiaoxuanPKU, I have rebased the PR so this should be good to go 🙇

**Comment by simon-mo** [comment]
nice! i don't even though this schema existed in the first place

File: tests/entrypoints/test_openai_server.py [hunk 4]
```diff
@@ -687,6 +687,22 @@ async def test_response_format_json_object(server, client: openai.AsyncOpenAI):
     assert loaded == {"result": 2}, loaded
 
 
+async def test_complex_message_content(server, client: openai.AsyncOpenAI):
+    resp = await client.chat.completions.create(model=MODEL_NAME,
```
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:691]: let's set temperature to 0 and seed so the output is determinstically 2

File: vllm/entrypoints/openai/protocol.py [hunk 5]
```diff
@@ -59,10 +62,32 @@ class ResponseFormat(BaseModel):
...
+    return value
+
+
 class ChatCompletionRequest(BaseModel):
     # Ordered by official OpenAI API documentation
     # https://platform.openai.com/docs/api-reference/chat/create
-    messages: List[Dict[str, str]]
+    messages: List[Dict[str,
+                        Annotated[str,
+                                  BeforeValidator(try_flatten_text_content)]]]
```
Comment (simon-mo) [inline vllm/entrypoints/openai/protocol.py:90]: can we just do a type union please?

**Review by simon-mo — CHANGES_REQUESTED** [review]
i actually think the proper fix would be support this in serving_completions directly to unwrap the text if complex schema is provided. we just added llava support and i think someone can add images as a follow up.

**Comment by fgreinacher** [comment]
> i actually think the proper fix would be support this in serving_completions directly to unwrap the text if complex schema is provided. we just added llava support and i think someone can add images as a follow up.

I had the same idea at first, but then I realized that this is not support by the transformers library (https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_chat.py#L55-L58). My thinking was that passing the complex format on therefore does not bring much benefit and just weakens typing. Full support could for sure be implemented, but would need quite some effort. Maybe we can start with the simple solution and evolve as needed?

File: vllm/entrypoints/openai/protocol.py (hunk 5 above)
Comment (fgreinacher) [reply vllm/entrypoints/openai/protocol.py]: Done

File: tests/entrypoints/test_openai_server.py (hunk 4 above)
Comment (fgreinacher) [reply tests/entrypoints/test_openai_server.py]: Good suggestion, done!

**Comment by DarkLight1337** [comment]
I have added support for single image input in #3978. However at this stage, VLM config is required when complex object is passed even when only text is passed.

File: vllm/entrypoints/openai/protocol.py [hunk 6]
```diff
@@ -59,10 +62,34 @@ class ResponseFormat(BaseModel):
...
+        # cases where just a single text content is passed in a message.
+        with contextlib.suppress(ValidationError):
+            single_text_content_list = TypeAdapter(
+                SingleTextContentList).validate_python(value)
+            return single_text_content_list[0].text
+    return value
+
+
+ChatCompletionMessage = Dict[str, Annotated[
+    str, BeforeValidator(try_flatten_text_content)]]
```
Comment (simon-mo) [inline vllm/entrypoints/openai/protocol.py:86]: still a bit unsatisfied that we cannot just do Union[str, List[TextContent]] and then turn the `TextContent` into `str` in the post processing in `ChatCompletionRequest`.

**Review by simon-mo — APPROVED** [review]
can you help do a final check on what the auto api documentation will show? it should be at http://localhost:8000/docs

File: vllm/entrypoints/openai/protocol.py (hunk 6 above)
Comment (fgreinacher) [reply vllm/entrypoints/openai/protocol.py]: OK, I have moved the flattening to a dedicated method. Let me know what you think @simon-mo 🙇

File: vllm/entrypoints/openai/protocol.py [hunk 7]
```diff
@@ -187,6 +198,14 @@ def logit_bias_logits_processor(
             logits_processors=logits_processors,
         )
 
+    @computed_field
+    @property
+    def normalized_messages(self) -> List[Dict[str, str]]:
+        return [{
+            key: value if isinstance(value, str) else value[0].text
+            for key, value in message.items()
+        } for message in self.messages]
```
Comment (fgreinacher) [inline vllm/entrypoints/openai/protocol.py:207]: I decided to keep this here so that we can benefit from Pydantic convenience.

**Comment by DarkLight1337** [comment]
#4355 uses the official type definitions from the `openai` Python library. This ensures consistency with using `openai.Client` to access the server.

**Comment by fgreinacher** [comment]
> Just a heads-up that #4355 uses the official type definitions from the `openai` Python library. This ensures consistency with using `openai.Client` to access the server. I think there is no need to maintain our own type definitions for the `message` inputs.

Oh, that's very nice. Also great that you already provide a placeholder for the changes in this MR :)

I'll put this MR back to Draft and will refactor once https://github.com/vllm-project/vllm/pull/4355 is merged.

**Comment by fgreinacher** [comment]
> > Just a heads-up that #4355 uses the official type definitions from the `openai` Python library. This ensures consistency with using `openai.Client` to access the server. I think there is no need to maintain our own type definitions for the `message` inputs.
> 
> Oh, that's very nice @DarkLight1337. Also great that you already provide a placeholder for the changes in this MR :)
> 
> I'll put this MR back to Draft and will refactor once #4355 is merged.

I have rebased and adapted the code, making this one much smaller, thanks @DarkLight1337 👍

**Comment by DarkLight1337** [comment]
> I have rebased and adapted the code, making this one much smaller, thanks @DarkLight1337 👍
> 
> @simon-mo This is ready for another round I'd say. The Neuron test is failing, but I don't think it's related to the changes in this PR.

Glad to help! I think you can take more parts from #4200 and enable multiple text inputs per message (concatenating them with a newline).

**Comment by fgreinacher** [comment]
> > I have rebased and adapted the code, making this one much smaller, thanks @DarkLight1337 👍
> > @simon-mo This is ready for another round I'd say. The Neuron test is failing, but I don't think it's related to the changes in this PR.
> 
> Glad to help! I think you can take more parts from #4200 and enable multiple text inputs per message (concatenating them with a newline). That way, we get to test out this functionality a bit more before extending it to images.

Wonderful, I took the relevant part and integrated it here 🙇

**Review by simon-mo — APPROVED** [review]
